### DISCUSSION TITLE
Proposal: out patterns, variable reassignment in pattern matching

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#6199] [Language Ideas]
One limitation of pattern matching is that variable patterns may only declare a new variable in scope.  If you want to assign that value to an existing variable you need to match it via a variable pattern into a new variable and then have a separate assignment afterwards:

```cs
string name;
if (obj is Person { Name: var n })
{
    name = n;
    // use name here
}
```

I think it would be convenient to have a pattern that allows you to extract a value directly into an existing variable:

```cs
string name;
if (obj is Person { Name: out name })
{
    // use name here
}
```

Unlike `var` this pattern cannot match any value, it would implicitly behaves as a type pattern of the same type as the variable.  So the following example would not match:

```cs
object obj = 123;
string name;

// does not match as the value is not compatible with string
if (obj is out name) {
}
```

Thoughts?

**Update:** It's been mentioned that the LDM has already considered this and was concerned about confusion as to the value of the variable in the case that the pattern did not match.  I see three possible outcomes in this case:

1. The variable retains whatever value it had before the pattern.  If the variable was not definitely assigned it remains not definitely assigned.
2. The variable is reset to `default`.  If the variable was not definitely assigned it remains not definitely assigned.
    * This would behave similarly to `out` parameters today, except that `out` parameters are always considered definitely assigned by the method.
3. The variable is reset to not definitely assigned in the `else` path.

**Comment by iam3yal** [comment]
Might be good idea to expand this proposal so the following would be possible too.
```csharp
if (obj is Person { Name: var out name })
{
    // use name here
}
```

**Reply by HaloFour** [reply]
> Might be good idea to expand this proposal so the following would be possible too.

What would that do differently than `var name` or `out name`?

**Reply by iam3yal** [reply]
@HaloFour I might be misunderstanding your proposal but with your proposal you have to declare a variable outside to the if statement still, right? and using `var name` in the pattern is not usable outside to the scope of the `if` statement so doing `var out name` does exactly what you're proposing but without having to declare the variable outside the `if` statement. I hope it's clear.

**Reply by svick** [reply]
@eyalalonn With `if (obj is Person { Name: var name })`, `name` is not usable outside of the `if`, because it is not definitely assigned. If `obj` is not a `Person`, what would be the value of `name`?

**Reply by iam3yal** [reply]
@svick The value would be `default` so null or whatever was assigned to it exactly like in the OP.

It's exactly like doing the following without the need to declare the variable `name` outside.
```csharp
string name;
if (obj is Person { Name: var n })
{
    name = n;
    // use name here
}
```

**Reply by HaloFour** [reply]
@eyalalonn 

> It's exactly like doing the following without the need to declare the variable `name` outside.

That's what variable patterns with `var` already does:

```cs
if (obj is Person { Name: var name })
{
    // use name here
}
```

Except that this does not assign a value to `name` if the pattern failed to match and the variable remains unassigned outside of the `if` block.

**Reply by iam3yal** [reply]
@HaloFour What I'm getting at is you can't provide initial value to `name` when you do `var name` because it isn't declared before the `if` statement.

**Reply by HaloFour** [reply]
@eyalalonn 

> What I'm getting at is you can't provide initial value to `name` when you do `var name` because it isn't declared before the `if` statement.

I don't get what that means.  You can't assign a value to a variable before it has been assigned, and `out var` is a declaration, so you still wouldn't be able to assign a value to the variable before/outside of the `if` block.

**Reply by iam3yal** [reply]
@HaloFour But in your example you can do the following:
```csharp
string name = "Anonymous";
if (obj is Person { Name: out name })
{
    // use name here
}
```
So what I'm suggesting is exactly the same as the following:
```csharp
name = "Anonymous";
if (obj is Person { Name: var out name })
{
    // use name here
}
```

**Reply by HaloFour** [reply]
@eyalalonn 

> So what I'm suggesting is exactly the same as the following:

I very much don't like that idea.  C# never permits referencing a variable before it has been declared and there's nothing about this proposal that I think should change that.

I'd suggest that variable "hoisting" should be a separate discussion.

**Reply by FaustVX** [reply]
@eyalalonn 
> @HaloFour But in your example you can do the following:
> 
> ```cs
> string name = "Anonymous";
> if (obj is Person { Name: out name })
> {
>     // use name here
> }
> ```
> 
> So what I'm suggesting is exactly the same as the following:
> 
> ```cs
> name = "Anonymous";
> if (obj is Person { Name: var out name })
> {
>     // use name here
> }
> ```

What's the point of writing `out var` instead of simply `out` here ?

**Reply by iam3yal** [reply]
@FaustVX With all due respect I think we've been through this, I thought it can be a good idea but turns out it can be confusing.

**Comment by HaloFour** [comment]
I've updated the proposal with some questions regarding what the compiler should do with the existing variable if the pattern doesn't match.

> **Update:** It's been mentioned that the LDM has already considered this and was concerned about confusion as to the value of the variable in the case that the pattern did not match.  I see three possible outcomes in this case:
> 
> 1. The variable retains whatever value it had before the pattern.  If the variable was not definitely assigned it remains not definitely assigned.
> 2. The variable is reset to `default`.  If the variable was not definitely assigned it remains not definitely assigned.
>     * This would behave similarly to `out` parameters today, except that `out` parameters are always considered definitely assigned by the method.
> 3. The variable is reset to not definitely assigned in the `else` path.

I'm partial to 1 myself, although I could see value in 2 and having parity with `Try*` methods even though I've heard before that it's confusing that the variable is overwritten when the method returns `false`.

**Reply by HaloFour** [reply]
Examples:

1.
    ```cs
    string name = "default";
    if (obj is Person { Name: out name }) {
        // ...
    }
    else {
        Debug.Assert(name == "default");
    }
    ```
2.
    ```cs
    string name = "default";
    if (obj is Person { Name: out name }) {
        // ...
    }
    else {
        Debug.Assert(name == default(string));
    }
    ```
3.
    ```cs
    string name = "default";
    if (obj is Person { Name: out name }) {
        // ...
    }
    else {
        // error CS0165: Use of unassigned local variable 'name'
        Console.WriteLine(name);
    }
    ```

**Reply by iam3yal** [reply]
I get your point now, thank you.

**Comment by RikkiGibson** [comment]
I think (1) is the right behavior here. In my view it is completely consistent with `out` arguments to methods. I would prefer not to do (3) because it would make this the first way that an expression can change a variable from *assigned* to *unassigned*. That will increase the computational complexity of the flow analysis.

Let's consider the `var name` scenario again.

```cs
void M0(object obj)
{
    if (obj is Person { Name: var name })
    {
        name.ToString(); // ok
    }
    name.ToString(); // error
}
```

The reason `name` is not definitely assigned at the `// error` above is because the pattern may not match. Compare to a scenario that uses a pattern which always matches:

```cs
struct Person
{
    public string Name { get; set; }
}

void M1(Person obj)
{
    if (obj is { Name: var name }) // pattern always matches
    {
        name.ToString(); // ok
    }
    name.ToString(); // ok
}
```

Similarly, when calls using `out` arguments are preceded by conditional tests, the variable is only treated as assigned in some code paths:

```cs
void M0_NonPattern(object obj)
{
    string name;
    if (obj is Person && Assign(out name, ((Person)obj).Name))
    {
        name.ToString(); // ok
    }

    name.ToString(); // error
}

void M1_NonPattern(Person obj)
{
    string name;
    if (Assign(out name, obj.Name))
    {
        name.ToString(); // ok
    }

    name.ToString(); // ok
}

bool Assign<T>(out T variable, T value)
{
    variable = value;
    return true;
}
```

There are a few other things which I think should be considered in the proposal:

## Timing of effects

Interestingly, even `var` assignments in patterns which are guaranteed to occur in the Roslyn implementation are still treated as conditional if the pattern may fail at another point in its evaluation. This is probably to push users toward treating patterns as having an undefined execution order.

```cs
void M3(Person obj)
{
    if (obj is { Name: var name, LastName: string lastName })
    {
        name.ToString(); // ok
        lastName.ToString(); // ok
    }

    name.ToString(); // error, but is always assigned in practice
    lastName.ToString(); // error
}

struct Person
{
    public string Name { get; set; }
    public string LastName { get; set; }
}
```

This raises a question of whether to delay the side-effects of `out` patterns until the pattern is known to be a match.

```cs
void M4()
{
    string name = "a";
    Person obj = new Person { Name = "b", LastName = null };
    if (obj is { Name: out name, LastName: string lastName })
    {
        name.ToString(); // ok
        lastName.ToString(); // ok
    }
    else
    {
        Console.Write(name); // write "a" or "b"? is it even defined by the language which one it is?
    }
}
```

I think I'm ok with saying that whether `name` is assigned by the pattern above is not defined by the language, and reflecting this in the flow analysis, while having it assign `name` in practice. I'd like to know how others feel, though.

## Conversions

It feels like it should be permitted to apply implicit conversions like we do for normal assignments.

```cs
struct Person
{
    int Age { get; set; }
}
void M5(Person obj)
{
    long age = obj.Age; // ok, implicit conversion from int to long
    if (obj is { Age: out age }) // ok
    {
    }
}
```

**Reply by HaloFour** [reply]
@RikkiGibson 

> I think (1) is the right behavior here. In my view it is completely consistent with `out` arguments to methods.

Actually that would be (2), methods are always required to set some value to `out` parameters, so typically they `default` them:

```cs
int i = 123;
string s = "foo";
if (int.TryParse(s, out i)) {
    // ..
}
else {
    Console.WriteLine(i); // prints 0
}
```

>  I would prefer not to do (3) because it would make this the first way that an expression can change a variable from _assigned_ to _unassigned_.

I agree, computation complexity aside I think it would also be a bit surprising as there's no way to "unassign" a variable in the language today.

> It feels like it should be permitted to apply implicit conversions like we do for normal assignments.

This seems trickier as conversions are not permitted for either `out` parameters or type patterns today:

```cs
int i = 123;
if (i is long l1) { } // error CS8121: An expression of type 'int' cannot be handled by a pattern of type 'long'.

object o = i;
if (o is long l2) { } // compiles but does not match

long l3;
int.TryParse("foo", out l3); // error CS1503: Argument 2: cannot convert from 'out long' to 'out int'
```

For the sake of consistency I think conversions should not be permitted and that the `out` pattern should be treated as if it were a type pattern of the same type as the declared variable.

**Reply by RikkiGibson** [reply]
No, I'm not suggesting (2). The out variable is still only assigned when the pattern matches--it's just that certain patterns happen to be guaranteed to match. There's no placeholder value that gets assigned when the pattern doesn't match, for example.

The point about conversions seems reasonable to me. Thanks!

**Reply by RikkiGibson** [reply]
> This seems trickier as conversions are not permitted for either `out` parameters or type patterns today

Revisiting this in a little more detail. It seems like certain conversions *are* permitted for type patterns. We'll just want to review the design to determine which conversions those are. For example:

```cs
class C
{
    public string Prop { get; set; }

    static void M(C c)
    {
        if (c is { Prop: object obj }) { }
    }
}
```

It's possible we would want to be *more* strict here for `out` patterns. But it's not obvious to me that would be the right thing.

**Reply by HaloFour** [reply]
I think upcasts are fine, I don't even really consider them "conversions".  This already works with patterns, and you can certainly assign a `string` to an `out object` parameter, the resulting variable must also be `object` but that would be the same with the pattern.

**Comment by jnm2** [comment]


**Reply by HaloFour** [reply]
@jnm2 

> I reread the definite assignment part at the top and now I'm not sure any proposed option would actually allow this.

Probably not, at least not without first assigning `variable` to some value:

```cs
protected override bool HasSingleVariable(
    LocalDeclarationStatementSyntax localDeclarationStatement,
    [NotNullWhen(true)] out VariableDeclaratorSyntax? variable)
{
    variable = null;
    return localDeclarationStatement.Declaration.Variables is [out variable];
}
```

Would be nice if that could be a one-liner or an expression-bodied member but I think considering the variable to be definitely assigned if the pattern doesn't match would be dangerous.

**Reply by jnm2** [reply]
The nice part is that many instances at my day job would still work with `is [out foo]`, due to a previous definite assignment.

**Comment by MgSam** [comment]
I think this proposal neatly addresses probably one of the more annoying limitations in pattern matching.

**Reply by FaustVX** [reply]
Yes.
This, and the ability to match against non-contant value.
Maybe with active pattern (#277), this could be achieved
